docs/intro/advanced-tutorial.rst

```diff
@@ -0,0 +1,165 @@
+
+        async def to_item(self):
+            item = {
+                "url": self.url,
+                "name": self.css("#main h3.name ::text").get(),
+                "product_id": self.css("#product ::attr(product-id)").get(),
+            }
+
+            # Simulates clicking on a button that says "View All Images"
+            response: web_poet.HttpResponse = await self.http_client.get(
+                f"https://api.example.com/v2/images?id={item['product_id']}"
+            )
+            item["images"] = response.css(".product-images img::attr(src)").getall()
+            return item
+
+
+It can be directly used inside the spider as:
+
+.. code-block:: python
+
+    import scrapy
+
+
+    def ProductSpider(scrapy.Spider):
+
+        custom_settings = {
+            "DOWNLOADER_MIDDLEWARES": {
+                "scrapy_poet.InjectionMiddleware": 543,
+            }
+        }
+
+        start_urls = [
+            "https://example.com/category/product/item?id=123",
+            "https://example.com/category/product/item?id=989",
+        ]
+
+        async def parse(self, response, page: ProductPage):
+            return await page.to_item()
+
+Note that we needed to update the ``parse()`` method to be an ``async`` method,
+since the ``to_item()`` method of the Page Object we're using is an ``async``
+method as well.
+
+This is also the primary reason why **scrapy-poet** requires ``scrapy>=2.6.0``
+since it's the minimum version that has full :mod:`asyncio` support.
+
+
+Meta
+====
+
+Using ``web_poet.Meta`` allows the Scrapy spider to pass any arbitrary information
+into the Page Object.
+
+Suppose we update the earlier Page Object to control the additional request.
+This basically acts as a switch to update the behavior of the Page Object:
+
+.. code-block:: python
+
+    import attr
+    import web_poet
+
+
+    @attr.define
+    class ProductPage(web_poet.ItemWebPage):
+        http_client: web_poet.HttpClient
+        meta: web_poet.Meta
+
+        async def to_item(self):
+            item = {
+                "url": self.url,
+                "name": self.css("#main h3.name ::text").get(),
+                "product_id": self.css("#product ::attr(product-id)").get(),
+            }
+
+            # Simulates clicking on a button that says "View All Images"
+            if self.meta.get("enable_extracting_all_images")
+                response: web_poet.HttpResponse = await self.http_client.get(
+                    f"https://api.example.com/v2/images?id={item['product_id']}"
+                )
+                item["images"] = response.css(".product-images img::attr(src)").getall()
+
+            return item
+
+Passing the ``enable_extracting_all_images`` meta value from the spider into
+the Page Object can be achieved by using **Scrapy's** ``Request.meta`` attribute.
+Specifically, any ``dict`` value inside the ``po_args`` parameter inside
+**Scrapy's** ``Request.meta`` will be passed into ``web_poet.Meta``.
```

Comment:
I think the fact that @BurnzZ you felt a need to call a key for web_poet Meta "``po_args``" means we should think about web_poet.Meta name again. Ideally, names should match.

Reply:
Hmmm, yes, a good point.

One option is to rename the `Meta` class as `PoArgs` to semantically match `request.meta["po_args"]`.

Maybe other names could be `PageArgs`, `PageParams`, and `PoParams`.

Reply:
This is something to change in web-poet itself, if we change it at all, right?

We could also use `request.meta["po_meta"]`. I don’t see a big issue with it, I think the repeated `meta` is acceptable in this scenario. We could use `request.meta["po_meta"]`, but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata.

Reply:
> This is something to change in web-poet itself, if we change it at all, right?

Yes, [this one](https://github.com/scrapinghub/scrapy-poet/pull/62/files#diff-a38b072945ddeae120d2462383b70ddc0f40967b53c849c49d6a3cbdf1917f53R217-R225) in particular, which is at its core.

> but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata

I think it's safe to say that it will be needed, for cases [like this](https://github.com/scrapinghub/scrapy-poet/pull/62/files#diff-261312b8086ac4bda934c1858d1479ee18c82ba2c22615f1f0ccc798a0dc30b4R162).

Reply:
> We could use `request.meta["po_meta"]`, but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata.

I made a typo here :facepalm: . I meant `"po"`. We could use `"po"` if we are sure po-related stuff in `request.meta` will always be limited to `po.Meta` data. Which I would not bet on, even if it seems so now.

Reply:
hey! Before we close it off, let's do a final check on the argument/dependency name. It affects both web-poet and scrapy-poet.

I think two main options are:

1. keep web_poet.Meta. This is similar to Scrapy's request.meta, so it can be familiar for developers who worked with Scrapy (?).
2. rename it to something like web_poet.Arguments, web_poet.PageParams, etc. 

After giving it some time to settle, I think I slighly (+0.5) prefer the second option, because of these reasons:

1. In most examples we provide, this argument ("po_meta") is used to control Page behavior. It's some additional per-request parameters, arguments we pass to a page object, which is specific to this page object. So, something like "arguments" or "params" make it more explicit.
2. In general, "meta" or "metadata" is some information about other information - about request in our case. Common examples of metadata are "date created", "author", etc. It's commonly used to aggregate information later, or to provide some additional information about a data point. I'm not sure how common is it to use "meta" or "metadata" to describe arguments which control the behavior; we don't call function parameters "metadata". In Scrapy this happens, but even there meta is used to provide after-the-fact information about the request as well (e.g. download_latency). Also, some uses of Scrapy's meta are deprecated in favor of cb_kwargs.
3. Naming it "meta" could lead to confusion with Scrapy's request.meta. First, user might expect that by depending on web_poet.Meta you'd get access to request.meta, which is not the case. Second, request.meta is mutable - for components like middlewares it can be used both as an input (to control the behavior), and as an output (to provide some information about what happened), while web_poet.Meta is not intended to be mutable, it's only there to be read. So, even for Scrapy users not having "Meta" in the name could allow to reduce the possible confusion.

----

Unrealted to the above, I think we shouldn't use "po" shortcut. We moved away from it (e.g. scrapy-poet used to be called scrapy-po), because it doesn't always sound great in English, some people were saying it like "poo" :)

Reply:
You've raised great points @kmike.

I'm inclined to rename `web_poet.Meta` into `web_poet.PageParams`. From the naming semantics, it's quite clear that it contains parameters for a page. In addition, perhaps we could also rename `request.meta["po_args"]` into `request.meta["page_params"]` for symmetry.